src/datahike/schema.cljc

```diff
@@ -0,0 +1,136 @@
+                                   :db/index {:db/valueType   :db.type/boolean
+                                              :db/unique      :db.unique/identity
+                                              :db/cardinality :db.cardinality/one}
+                                   :db/unique {:db/valueType   :db.type/unique
+                                               :db/unique      :db.unique/identity
+                                               :db/cardinality :db.cardinality/one}
+                                   :db/isComponent {:db/valueType :db.type/boolean
+                                                    :db/unique :db.unique/identity
+                                                    :db/cardinality :db.cardinality/one}
+                                   :db/txInstant {:db/valueType :db.type/instant
+                                                  :db/unique :db.unique/identity
+                                                  :db/index true
+                                                  :db/cardinality :db.cardinality/one}
+                                   :db/retracted {:db/valueType :db.type/long
+                                                  :db/unique :db.unique/identity
+                                                  :db/cardinality :db.cardinality/many}
+                                   :db/doc {:db/valueType :db.type/string
+                                            :db/index true
+                                            :db/cardinality :db.cardinality/one}
+                                   :db.install/_attribute {:db/valueType   :db.type.install/_attribute
+                                                           :db/unique      :db.unique/identity
+                                                           :db/cardinality :db.cardinality/one}})
+
+(def schema-keys #{:db/ident :db/isComponent :db/valueType :db/cardinality :db/unique :db/index :db.install/_attribute :db/doc})
+(def meta-keys #{:db/txInstant :db/retracted})
+
+(defn meta-attr? [attr]
+  (s/valid? ::meta-attribute attr))
+
+(defn meta-entity? [entity]
+  (some #(contains? entity %) meta-keys))
+
+(defn schema-attr? [attr]
+  (s/valid? ::schema-attribute attr))
+
+(defn value-valid? [[_ _ a v _] schema]
+  (let [schema (if (or (meta-attr? a) (schema-attr? a))
+                 implicit-schema-spec
+                 schema)
+        value-type (get-in schema [a :db/valueType])]
+    (s/valid? value-type v)))
+
+(defn instant? [^Datom datom schema]
+  (let [a (.-a datom)
+        schema (if (or (meta-attr? a) (schema-attr? a))
+                 implicit-schema-spec
+                 schema)]
+    (= (get-in schema [a :db/valueType]) :db.type/instant)))
+
+(defn schema-entity? [entity]
+  (some #(contains? entity %) schema-keys))
+
+(defn schema? [schema]
+  (s/valid? ::schema schema))
+
+(defn describe-type [schema-type]
+  (s/describe schema-type))
+
+(defn find-invalid-schema-updates [entity attr-schema]
```

Comment:
Why are these proper rules for schema upgrades? I assume it is following some datomic rational. It is important that downstream queries do not break after schema changes and it can be the case with datalog that allowing multiple values for an attribute makes a query return less (e.g. nothing if it checks for uniqueness itself). We do not strictly have to have a good schema-update strategy in place, because it is actually context-specific on how the database is read, which cannot be known in general, but this is something we at least need to document well.

Reply:
I will create some documents about database migrations in general.

Reply:
Fixed with #70.